enhancements/ingress/set-delete-http-headers.md

```diff
@@ -0,0 +1,1031 @@
+respectively, for both the `Set` and `Delete` actions enables the template to preserve ordering.
+
+This feature's implementation will be similar to an existing feature called logging headers which was implemented in
+[router](https://github.com/openshift/router/commit/40ee0ba472a4ac503f1e6c97d4464f559101c97c) and 
+[cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/commit/c5139cb07c3a99a8a3b18a6c9321310add51b075).
+
+B) Setting or Deleting HTTP headers via Route's Spec by Users.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/router
+* openshift/openshift-apiserver
+
+The router configures HAProxy using a configuration template. The template is populated with data from `ServiceAliasConfig` values, where each `ServiceAliasConfig` represents a Route.
+
+The enhancement adds two fields in `ServiceAliasConfig`: `HTTPResponseHeaders`, `HTTPRequestHeaders` for request or response headers to be set or deleted in the
+the `haproxy.config` sections corresponding to the following backends: plain http backend, backend with TLS terminated at the edge, secure backend with re-encryption ([code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L609)).
+
+The router sets this variable based on the Route's `spec.httpHeaders.actions.response[*].action.set` or `spec.httpHeaders.actions.response[*].action.type` or `spec.httpHeaders.actions.request[*].action.set` or `spec.httpHeaders.actions.request[*].action.type` field values.
+`ServiceAliasConfig` is populated by [the router's `createServiceAliasConfig` method](https://github.com/openshift/router/blob/3065f6583f3925328fbdbfe95e3bc7bb7a084d33/pkg/router/template/router.go#L908-L935)
+
+```go
+// ServiceAliasConfig is a route for a service.  Uniquely identified by host + path.
+type ServiceAliasConfig struct {
+    ...
+    // For route based setting of http response headers
+    HTTPResponseHeaders []HTTPHeader
+
+    // For route based setting of http request headers
+    HTTPRequestHeaders []HTTPHeader
+}
+```
+```go
+// HTTPHeader specifies an HTTP header that should be set/replace/append
+type HTTPHeader struct {
+// Name specifies an HTTP header name.
+Name string
+
+// Value specifies the header value.
+Value string
+
+// Action specifies the action to be performed.
+Action string
+}
+```
+
+In this case the ordering of actions will be maintained as we will be having access to the RouteSpec inside the router. So, we will run a loop
+on either `spec.httpHeaders.actions.response[*].action.set` or `spec.httpHeaders.actions.response[*].action.type` or `spec.httpHeaders.actions.request[*].action.set` or `spec.httpHeaders.actions.request[*].action.type` which will have the desired
+sequence. 
+
+If request and response headers are provided in the API then first we run loop through request headers and then response headers.
```

Comment:
I think this says the same thing, a little more clearly.
```suggestion
The sequence of action operations will be maintained according to the router RouteSpec. First, we iterate  over `spec.httpHeaders.actions.request[*].action.set` or `spec.httpHeaders.actions.request[*].action.type`, then over `spec.httpHeaders.actions.response[*].action.set` or `spec.httpHeaders.actions.response[*].action.type`.
```

Reply:
done